// gatsby-plugin-s3/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var path_1 = __importDefault(require("path"));
exports.CACHE_FILES = {
    config: path_1.default.join('.cache', 's3.config.json'),
    params: path_1.default.join('.cache', 's3.params.json'),
    routingRules: path_1.default.join('.cache', 's3.routingRules.json'),
};
exports.DEFAULT_OPTIONS = {
    bucketName: '',
    params: {},
    mergeCachingParams: true,
    generateRoutingRules: true,
    generateIndexPageForRedirect: true,
    generateMatchPathRewrites: true,
    removeNonexistentObjects: true
};
exports.CACHING_PARAMS = {
    '**/**.html': {
        CacheControl: 'public, max-age=0, must-revalidate'
    },
    'static/**': {
        CacheControl: 'public, max-age=31536000, immutable',
    },
    '**/**/!(sw).js': {
        CacheControl: 'public, max-age=31536000, immutable'
    },
    '**/**.css': {
        CacheControl: 'public, max-age=31536000, immutable'
    },
    'sw.js': {
        CacheControl: 'public, max-age=0, must-revalidate'
    }
};
